tests/testsuite/git_auth.rs

```diff
@@ -137,7 +137,7 @@ fn http_auth_offered() {
         .with_stderr_contains(&format!(
             "\
 [UPDATING] git repository `http://{addr}/foo/bar`
-[ERROR] failed to get `bar` as a dependency of package `foo v0.0.1 [..]`
+[ERROR] failed to prefetch dependencies
```

Comment:
This looks like it's losing the "as a dependency of..." context, is there a way to preserve that?

Reply:
I thought a bit about this, and I think this actually isn't a great loss. The fidelity is specifically _only_ lost if a given source is inaccessible, in which case _any_ crate from that source is unavailable. If a given crate isn't available or something like that, the old verbose error will still be printed, since it will only become apparent once we start trying to `load` or `download` that dependency.

The deeper reason why the error was "dumbed down" here is that we are bulk fetching all dependencies of all packages in the workspace. If the source for one of those dependencies errors, all we (as in, the caller that prints this message) knows is that a source for _some_ dependency failed. After all, we just handed over a bunch of dependencies all at once. That said, the actual package we failed to fetch is apparent in the "Cause by" line just below this error:
https://github.com/rust-lang/cargo/blob/cd5281d857b896714e06c4fc237267804f04df19/tests/testsuite/git_auth.rs#L140-L143
So the only thing we've really lost is the connection to which crate in the workspace had that dependency in the first place. We could remedy that by passing in the name of the package that incurred each dependency to `prefetch`, thread that information all the way down through the registry code to where it discovers that the source is unavailable, and then surface it there, but that's a decent amount of added complexity for (arguably) little gain.

Reply:
This is now fixed. Or rather, worked around. We will now only give errors that mention prefetching if `-Z http-registry` is passed. And in the future, when using the real solver for all fetching, the errors will be "good" even with `-Z http-registry`.